File: 2762-continuous-subarrays/2762-continuous-subarrays.cpp
class Solution {
public:
    long long continuousSubarrays(vector<int>& nums) {
        deque<int> max_q;
        deque<int> min_q;

        int left = 0;
        long long ans = 0;

        for(int right = 0 ; right < nums.size() ; right++){
            while(!max_q.empty() && max_q.back() < nums[right]){
                max_q.pop_back();
            }
            max_q.push_back(nums[right]);
            
            // Maintain min deque
            while (!min_q.empty() && min_q.back() > nums[right]) {
                min_q.pop_back();
            }
            min_q.push_back(nums[right]);
            
            // Shrink window if maximum element - minimum element > 2
            while (max_q.front() - min_q.front() > 2) {
                if (max_q.front() == nums[left]) {
                    max_q.pop_front();
                }
                if (min_q.front() == nums[left]) {
                    min_q.pop_front();
                }
                left++;
            }
            
            // Add number of valid continuous subarrays ending at 'right'
            ans += (right - left + 1);
        }

        return ans;
    }
};

File: 2762-continuous-subarrays/2762-continuous-subarrays.py
class Solution:
    def continuousSubarrays(self, nums: List[int]) -> int:
        left = 0
        ans = 0
        max_q = deque()
        min_q = deque()

        for right in range(len(nums)):
            while max_q and max_q[-1] < nums[right]:
                max_q.pop()
            max_q.append(nums[right])
        
            while min_q and min_q[-1] >nums[right]:
                min_q.pop()
            min_q.append(nums[right])

        
            while max_q[0] - min_q[0] > 2:
                if max_q[0] == nums[left]:
                    max_q.popleft()
                if min_q[0] == nums[left]:
                    min_q.popleft()
                
                left += 1
            
            ans += (right - left + 1)

        return ans